client/app/pages/queries/QuerySource.jsx

```diff
@@ -41,6 +42,98 @@ function chooseDataSourceId(dataSourceIds, availableDataSources) {
   return find(dataSourceIds, id => includes(availableDataSources, id)) || null;
 }
 
+function confirmOverwrite() {
+  return new Promise((resolve, reject) => {
+    Modal.confirm({
+      title: "Overwrite Query",
+      content: (
+        <React.Fragment>
+          <div className="m-b-5">It seems like the query has been modified by another user.</div>
+          <div>Are you sure you want to overwrite the query with your version?</div>
+        </React.Fragment>
+      ),
+      okText: "Overwrite",
+      okType: "danger",
+      onOk: () => {
+        resolve();
+      },
+      onCancel: () => {
+        reject();
+      },
+      maskClosable: true,
+      autoFocusButton: null,
+    });
+  });
+}
+
+function saveQuery(query, data, options) {
+  query = clone(query);
+  options = {
+    successMessage: "Query saved",
+    errorMessage: "Query could not be saved",
+    ...options,
+  };
+
+  if (isObject(data)) {
+    extend(query, data);
+    // Don't save new query with partial data
+    if (query.isNew()) {
+      return Promise.resolve(query);
+    }
+    data = { ...data, id: query.id, version: query.version };
+  } else {
+    data = pick(query, [
+      "id",
+      "version",
+      "schedule",
+      "query",
+      "description",
+      "name",
+      "data_source_id",
+      "options",
+      "latest_query_data_id",
+      "is_draft",
+    ]);
+  }
+
+  // omit pendingValue before saving
```

Comment:
```suggestion
  // omit parameter properties that don't need to be stored
```
(or whatever like this, I forgot to update this comment in the Parameter structure update PR :sweat_smile:)

Reply:
Done :+1: